flash_map:  generate content from DTS completely.

findlayfeng opened the pull request:
This patch add functionality for extract complete flash map content form DTS. This makes DTS proper interface for providing partitioning information without hard-coding flash areas of partitions.

Changes also allow to declare partition nodes inside embedded and external flash memory simultaneously.

Before this PR flash map configuration requires both customized flash map c-structure and preparation of DTS description.

TODO:

- [x] automatic generation of the list of flash devices from dts
- [x] automatic generation of the list of flash area from dts
- [x] align DFU support modules
  - [x] Mcuboot support
  - [x] Flash img support
  - [x] Update zephyr img mgmt port
  - [x] Update usb dfu
  - [x] Update dfu tests
- [x] patch for mcuboot itself https://github.com/JuulLabs-OSS/mcuboot/pull/405
- [x] patch for mcumgr upstream https://github.com/apache/mynewt-mcumgr/pull/20

findlayfeng commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -31,13 +31,10 @@ struct driver_map_entry {
 	const char * const name;
 };
 
+#define FLASH_DEV_FOO(i, _) {i, FLASH_DEV_## i ##_NAME} ,
```
Can't pass the format check here
I am based on the example of UTIL_LISTIFY

Commit:
dfu/mcuboot: use flash_map instead of flash API

Patch introduces flash_map subsystem to operate on flash
footprint instead of direct operation using flash_driver API.
Flash area ID is used in API instead of direct flash-bank-offsets.
Changes allows to support operation on the partition in any flash
device.

flash_map was not available when this subsystem was introduced.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
dfu/flash_img: use flash_map instead of flash API

Patch introduces flash_map subsystem to operate on flash
image instead of direct operation using flash_driver API.
Changes allows to support operation on the image in any flash
device.

flash_map was not available when this subsystem was introduced.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
ext/lib/mgmt/mcumgr: align zephyr port to changes in dfu sybsy'es

Changes in flash_map API makes flash_area structure proper
interface for point the image area instead of direct flash-bank-offsets.

This patch align code to changed APIa and allows to support operation
on the partition in any flash device.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
usb/class/usb_dfu: align to changes in dfu subsys'es

Changes in flash_map API makes flash_area structure proper
interface for point the image area instead of direct flash-bank-offsets.

This patch align code to changed APIa and allows to support operation
on the partition in any flash device.

Signed-off-by: Findlay Feng <i@fengch.me>

carlescufi commented:
@nvlsianpu can you take a look?

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -164,115 +162,123 @@ static const struct boot_swap_table boot_swap_tables[] = {
 
-	rc = boot_flash_write(offs, buf, sizeof(buf));
+	rc = flash_area_write(fa, offs, buf, sizeof(buf));
+	flash_area_close(fa);
```
honestly I don't care about closing flash areas in zephyr-code as it is concept form mynew, and now no one know whom it serve.

findlayfeng replied:
Are you saying that open and close do not need to be in pairs?

nvlsianpu replied:
yes, exactly that - you can decide to keep your code as is. (flash_are_cloce is nop )

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -361,40 +367,48 @@ static int boot_magic_code_check(const u32_t *magic)
 {
-	u32_t magic[4];
+	const struct flash_area *fa;
+	u32_t magic[(sizeof(u32_t) - 1 + BOOT_MAGIC_SZ) / sizeof(u32_t)];
```
why? add comment in the code.

findlayfeng replied:
BOOT_MAGIC_SZ is defined, I want it to be used instead of a fixed value 16

nvlsianpu replied:
I saw that.

nvlsianpu reviewed (changes requested):
Commit messages
dfu: mcuboot: : Support flash map
This commit modifies zephyr subsystem which is dedicated for interact with mcuboot, not mcuboot-itself. It introduces flash_map subsystem to operate on flash footprint instead of flash_driver. So, refine commit message.

nvlsianpu commented:
@findlayfeng I will provide more comments soon. Nice to see you are still active. Can you rebate this PR onto newest master (and resolve conflicts)? - so I will be able to test it wit newest master.

I have also the  question - whether this PR fulfill flowing objective?:
"provide information about flash drivers for each image slot (using flash_map subsystem as well). We want to run mcuboot and DFU along with external serial flash.
So far reference to the flash driver was hardcoded."

findlayfeng commented:
In fact, this is exactly what I am doing.
The work on mcuboot has been completed and has not been submitted because the new api has been referenced and the new api has not yet been merged.

nvlsianpu reviewed (changes requested):
commit message:
dfu: flash_img: ~~Support~~involvew flash map

Now  falsh img uses the falsh area id to determine which ~~disk~~flash partition~~ is
being operated.

fix typos as well

nvlsianpu commented:
>In fact, this is exactly what I am doing.
The work on mcuboot has been completed and has not been submitted because the new api has been referenced and the new api has not yet been merged.

Great - I was almost sure thet it does that.

nvlsianpu reviewed (changes requested):
commit:
ext: img_mgmt: update api about dfu

Normally proces said that code from ext folder should be added to appropriate external project 1st. In this case changes are atomic (zephyr changes immediately required mcumgr changes), so I would update external project once this PR will be merged. @carlescufi can we do it in such a manner?

nvlsianpu commented:
@carlescufi OK -  I would update external project once this PR changes will be approved. Then we can update this PR and merge it immediately. This will be a "transaction".

carlescufi commented:
@nvlsianpu sounds good. We should still rebase this first.

nvlsianpu reviewed (changes requested):
commit:
`tests: img_util: update api about dfu`

should be squashed to the commit which changes img_util.

nvlsianpu reviewed (changes requested):
commit:
tests: mcuboot: update api about dfu

should be squashed to the commit which modifies dfu/mcuboot subsytem.

nvlsianpu commented:
@findlayfeng any chance for update this PR?

findlayfeng commented:
@nvlsianpu Today is the last day of the [Mid-Autumn Festival](https://en.wikipedia.org/wiki/Mid-Autumn_Festival) holiday in China.
Rebase has been completed
I am going to sleep now, I will deal with this RP tomorrow.

nvlsianpu commented:
Thanks for rebase and fixes - I just have been starting to test how this works (well so far :D).

nvlsianpu commented:
@findlayfeng can you put a working example of external flash-controller declaration. I can't build mine with success.
here what I have put to overlay file for nrf52840_pca10056:
```
/ {
	flash-controller2@7001E000 {
			compatible = "turbo,turbo-flash-controller";
			reg = <0x7001E000 0x550>;

			#address-cells = <1>;
			#size-cells = <1>;

			label="EXT_FLASH_DRV_NAME";

			flash1: flash@0 {
				compatible = "serial-flash";
				label = "SERIAL_FLASH";
				erase-block-size = <4096>;
				write-block-size = <4>;
			};
	};
};

&flash1 {
	/*
	 * For more information, see:
	 * http://docs.zephyrproject.org/latest/devices/dts/flash_partitions.html
	 */
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		external_partition: partition@0 {
			label = "external";
			reg = <0x000000000 0x00002000>;
		};
	};
};
```

nvlsianpu commented on dts/bindings/mtd/serial-flash.yaml:
```diff
@@ -0,0 +1,19 @@
+      category: required
+      description: compatible strings
+      constraint: "serial-flash"
+
```
Thought that an external serial flash should have erase-block-size and write-block-size properties as well.

findlayfeng replied:
Zephyr's flash erase is based on FLASH_PAGE_LAYOUT.
Not erase-block-size, this value is actually used when some of the on-chip flash defines the controller.
The serial-flash PAGE_LAYOUT will not generate according to this value.
As for whether you want to use it in serial-flash
I don't think it is necessary to use
Because the spi-flash chip is external and may be replaced
And spi-flash has its own discovery protocol
You can find out about [my other PR](https://github.com/findlayfeng/zephyr/tree/fix_spi_flash)
So I allow the size of the area to be set to 0 to automatically adjust
For the same reason I deleted write-block-size in serial-flash

nvlsianpu replied:
Thanks for this elaboration reason behind unrepresented properties.

I agree that for some devices it is a case that the layout can and should be generate form information received by a serial protocol.

I prefer to by mor generic anyway. So some serial devices doesn't provide such information by any mean. So we need to have it hard-coded somewhere. The best place for it is DT as they are hardware properties. In most expected cases External flash is not removable - it is a constant part of embedded device.

Sumarized I would add this properties as optional.

findlayfeng replied:
This part has been updated

findlayfeng commented:
/ {
	model = "xxxx";
	compatible = "nordic,nrf51822";

	chosen {
		zephyr,flash = &flash0;
		zephyr,flash_map = &flash0, &flash1;
	};

	flash1: serial-flash@0 {
		compatible = "serial-flash";
		reg = <0 DT_FLASH_SIZE>;
		label = "FLASH_1";

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			slot1_partition: partition@0 {
				label = "image-1";
				reg = <0x00000000 0x30000>;
			};

			storage_partition: partition@30000 {
				label = "storage";
				reg = <0x00030000 0>;
			};
		};
	};

};

&flash0 {
	/*
	 * For more information, see:
	 * http://docs.zephyrproject.org/devices/dts/flash_partitions.html
	 */
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		boot_partition: partition@0 {
			label = "mcuboot";
			reg = <0x00000000 0xC000>;
		};
		slot0_partition: partition@C000 {
			label = "image-0";
			reg = <0x0000C000 0x30000>;
		};

		scratch_partition: partition@3c000 {
			label = "image-scratch";
			reg = <0x0003c000 0x4000>;
		};

	};
};

findlayfeng commented:
Important is
```
chosen {
		zephyr,flash = &flash0;
		zephyr,flash_map = &flash0, &flash1;
	};
```

If zephyr, flash_map has no value
Automatic use of flash
So the old configuration does not need to be modified

Here is just a reference to the device based on label
The definition of the controller I think is the change related to flash

nvlsianpu commented:
@findlayfeng 
I see now that this PR introduce outstanding functionality required for all external flash devices support (SPI, QSPI. I2C and parallel as well). There is nothing very specific for only on of these interfaces. So I recommend to change terms used elsewhere (as keyword etc.) in this PR from `SPI`/`Serial` to `External`.
for example:
 compatible = "~~serial~~external-flash";

cc @galak ^^

findlayfeng commented:
@nvlsianpu 
I don't think so, because the direction that follows should be a bus-based device configuration. Just like rtc
Based on this, I think the current description is appropriate.
Dependent on the partition tag instead of "serial-flash" or something else
If you want to introduce devices from other buses, I think you should increase the configuration.

nvlsianpu commented:
@findlayfeng So your thought is that this functionality is dedicated to seria flash devices, right?

erwango reviewed (commented):
Seems interesting but it would be great to have a way to test this for the review.

findlayfeng commented:
@nvlsianpu 
I mean you can add a dts configuration file like dts/bindings/mtd/serial-flash.yaml

nvlsianpu commented:
@erwango Use the last commit form my fork https://github.com/nvlsianpu/zephyr/tree/fix_flash_map_sandbox for generated flash_map test.

nvlsianpu commented:
> Seems interesting but it would be great to have a way to test this for the review.

@erwango 
Can you elaborate more - Did you think about example which lists each flash_areas with it's properties?

nvlsianpu commented on scripts/dts/extract/flash.py:
```diff
@@ -10,6 +10,11 @@
 
+flash_dev_label_prefix = {
+    "soc-nv-flash":"soc_flash",
+    "serial-flash":"spi_flash"
```
can you change to "serial-flash":"serial_flash" or another equivalent?

findlayfeng replied:
This array is for compatibility with old code
The old code uses the name spi flash

findlayfeng replied:
Modified to "serial-flash": "serial_flash" and directly delete "serial-flash": "spi_flish" will have the same effect

nvlsianpu replied:
What a old code? Can be modified as well?

> Modified to "serial-flash": "serial_flash" and directly delete "serial-flash": "spi_flish" will have the same effect

I don't understand what you want to say.

findlayfeng replied:
These two names were defined and used in the previous code.
If there are other places where these two definitions are used, I am in order to be compatible with the previous name.
Amount, this period of time comes in contact, the current zephyr code does not have to consider the name compatibility.
This paragraph can be deleted.

nvlsianpu reviewed (changes requested):
@findlayfeng Can you apply changes I request.  Unfortunately I didn't check commit message again.
After expected changes I will approve the patch and recommend it to be merged.

galak commented:
Is there an example of what the DT is suppose to look like?  Not sure I follow what's this is trying to do with generating a list of flash devices, etc.

findlayfeng commented:
@galak This PR is not generating a list of flash devices,
Is to give a list of flash devices and scan the defined partitions on these devices

nvlsianpu commented:
@findlayfeng I can send example of DT description requested by @galak onto your commits - It might help with further discussion. Can I?

findlayfeng commented on scripts/dts/extract/flash.py:
```diff
@@ -18,6 +23,76 @@ class DTFlash(DTDirective):
+            label = self.get_label_string([def_label, str(index), 'name'])
+            prop_def[label] = name
+
+            if dev_type in flash_dev_label_prefix:
```
This array is used like this

findlayfeng replied:
"serial-flash": "serial_flash" is equivalent to writing nothing.

findlayfeng commented on include/flash_map.h:
```diff
@@ -33,9 +33,6 @@ extern "C" {
 #include <stddef.h>
 #include <sys/types.h>
 
-#define SOC_FLASH_0_ID 0	/* device_id for SoC flash memory driver */
```
Here is the reason why this array is defined
In order to be compatible with the old name

findlayfeng reviewed (commented):
my English is poor
I directly pointed out the relevant code
@nvlsianpu

findlayfeng commented:
> n. Can I?

yep

findlayfeng commented:
> @findlayfeng I can send example of DT description requested by @galak onto your commits - It might help with further discussion. Can I?

yep

nvlsianpu commented:
@findlayfeng can you act for this PR resolution. For instance you can prepare a expected by this patch external memory DT description for a board (for instance nrf52840_pca10056 has SPI/qspi die on the pcb)?

findlayfeng commented:
@nvlsianpu 
The board I use is not a development board.
I need to prepare a board that is in the boards.

carlescufi commented:
@findlayfeng have you added QSPI commits here by mistake?

findlayfeng commented:
Is an example that relies on qspi's flash
@carlescufi

nvlsianpu commented:
@findlayfeng 
Thanks for adding the qspi example!
For sure it is required (by the contribution process itself) to isolate changes. So please remove most QSPI code form here and keep it exclusively in #11278.

For presentation of this PR contribution I would keep here only changes you added in files:
dts/arm/nordic/nrf52840.dtsi
dts/bindings/flash_controller/nordic,nrf-qspi-flash-controller.yaml
soc/arm/nordic_nrf/nrf52/dts_fixup.h

nvlsianpu commented on include/dfu/flash_img.h:
```diff
@@ -14,7 +14,7 @@ extern "C" {
 struct flash_img_context {
 	u8_t buf[CONFIG_IMG_BLOCK_BUF_SIZE];
-	struct device *dev;
+	const void *flash_area;
```
> In fact, the original code is problematic, it requires device.h to be written in front.
I think I should also add #include <zephyr/types.h>

I missed that. Add both includes and use flash_area pointer explicite.

findlayfeng commented:
@nvlsianpu 
Sorry, I am doing other work recently.
What do you mean by splitting PR so that they are no longer dependent on it?
My original idea was to make #11278 first merged.
If you remove the example, the first commit (64707cc) is also the common dependency of the two PRs.

findlayfeng commented:
@nvlsianpu Update example using an area of soc flash ^_^

nvlsianpu commented:
I removed commit `samples: flash_map: add sample for nrf52840_pca10056` as it was not my intention to ask about it.

nvlsianpu commented:
@lemrey Can you test this along with mcuboot? On pca10040 or pca10056. Looks like PR is ready I'm on vacation" so can't test this.

lemrey commented:
@findlayfeng @nvlsianpu there is an issue with the slots:
in `loader.c` : ` <wrn> mcuboot: Cannot upgrade: slots are not compatible`

findlayfeng commented:
Mcuboot needs some modifications

nvlsianpu commented:
Works firn for me. So strange.

@findlayfeng 
> Mcuboot needs some modifications

What changes?

nvlsianpu commented:
^^ we discovered than on my build machne we have got in <build-dir>zephyr/include/generated/generated_dts_board.h
```
#define FLASH_AREA_1		IMAGE_0
#define FLASH_AREA_2		IMAGE_1
```
while on @lemrey machine he has got
```
#define FLASH_AREA_0        IMAGE_0 
#define FLASH_AREA_4        IMAGE_1
```

Question is why this behaves diferently.

findlayfeng commented:
> ^^ we discovered than on my build machne we have got in zephyr/include/generated/generated_dts_board.h
> 
> ```
> #define FLASH_AREA_1		IMAGE_0
> #define FLASH_AREA_2		IMAGE_1
> ```
> while on @lemrey machine he has got
> 
> ```
> #define FLASH_AREA_0        IMAGE_0 
> #define FLASH_AREA_4        IMAGE_1
> ```
> Question is why this behaves diferently.

The numbers are assigned sequentially,
Decided on the script to find the order of this area,
Therefore, you should not use a fixed value when using it.

nvlsianpu commented:
@lemrey @carlescufi @nashif @findlayfeng @ioannisg  I tweaks this PR by edit commit messages. I tested it against mcuboot successfully as well. Also opened required PR to mcuboot and mcumgr up-steram.

Pleas review in your convenience.

lemrey commented:
@nvlsianpu please double-check serial recovery.

nvlsianpu commented:
@lemrey Thanks. I investigated and spot that mcuboot needs some more tweaks to be compatible with this patch. Need to rework [flash_area_sector_from_off](https://github.com/JuulLabs-OSS/mcuboot/blob/master/boot/zephyr/flash_map_extended.c#L60) function.

carlescufi commented:
@lemrey @nvlsianpu @erwango Is this ready go in?

galak commented on dts/bindings/mtd/spi-nv-flash.yaml:
```diff
@@ -0,0 +1,11 @@
+---
```
Where is this used?

nvlsianpu replied:
We want to use this for further spi flash. An example for that was in this PR before (something like https://github.com/zephyrproject-rtos/zephyr/pull/11278/files#diff-0ad92537b7c3e2041e8e2a6b9a30698b) - but it was tu much for this PR. So using of this will be the next step (subsequent PR)

galak replied:
Let's remove it for now, and keep this to things used by what's in the code base.

nvlsianpu commented:
@lemrey Can you send me your generated_dts_board.h file?

nvlsianpu commented:
^^ pleas also check whether disabling progressive erasing mitigates the issue.

lemrey commented:
The MCUBoot hex image generated on my machine works on @nvlsianpu and not mine. We suspect it is related to my setup, given that this also happens on master for me; it'd be great though if somebody else could check this.

lemrey reviewed (approved):
To clarify: I confirmed that the initial issue I had with this PR is fixed by the related PR to MCUboot's repo.
The issue with Serial Recovery, I _suspect_ being related to my setup and unrelated to this PR so this is good enough for me.

galak reviewed (changes requested):
What's the high level idea behind what we'd generate from the DTS?

nvlsianpu commented:
High level idea is to allow partitions in external flash storage and embedded simultaneously. This is step required by suport external memory for mcuboot for instance.

nashif commented:
flash_map test (tests/subsys/storage/flash_map) fails with this pr on nrf52840_pca10056

```
***** Booting Zephyr OS zephyr-v1.13.0-4010-ge4c6b5e790 *****
Running test suite test_flash_map
===================================================================
starting test - test_flash_area_to_sectors
***** USAGE FAULT *****
  Attempt to execute undefined instruction
***** Hardware exception *****
Current thread ID = 0x200045ec
Faulting instruction address = 0xc294
Fatal fault in ISR! Spinning...
```

nashif commented on subsys/storage/flash_map/flash_map_default.c:
```diff
@@ -8,54 +8,18 @@
 #include <zephyr.h>
```
The way this file was used before is not the same after this PR, this file AFAIK was the hardcoded flash map which could be replaced by a custom one using a Kconfig option. If I understand correctly, it is not needed now, the partitions can now be customised in DTS overlay and there is no need for a user to provide a hardcoded C based default map, am I correct here?

nvlsianpu replied:
Yes you are exactly correct - this file name is confusing so will change it.

nvlsianpu commented:
@nashif The bug was fixed
@galak:
Currently in zephyr the flash_map is static instance – hard-coded instance which present flash-partitions description from DT. In order to customize the flash_map need to provide C file with alternative flash_map structure, and (very likely) customize DTS as well. Flash_map entries are flash_area structures (fa). fa are identified by hard-coded ID.
Fa assignment to its flash devices is hard-coded by written by hand flash devices ID.
In DT It was impossible to generate partition information from more than one flash node.

For achieve automated generation of flash_map was required to do:
- manage fa IDs: now appropriate labels  (thanks to DT partition node name) and fa numeration are generated automatically
- manage relations between fa and its flash device. Flash devices Id are extracted from DT. Patch keep two kind of flash devices: SoC based and on serial interface. Fa assignment to a device is determined thanks to position in DTS structure.
- flash map customization by C file is removed as not needed anymore – now DTS become the best interface for flash_map customization.
- DFU support modules are switched to operate on any flash device thanks to using flash_map to operate on flash
footprint instead of direct operation using flash_driver API.
Flash area ID is used in API instead of direct flash-bank-offsets.

galak commented on scripts/dts/extract/flash.py:
```diff
@@ -10,6 +10,11 @@
 from extract.default import default
 from extract.reg import reg
 
+flash_dev_label_prefix = {
```
We should remove this, we don't need to generate it, just generate the flash controller label name for each area.

galak reviewed (dismissed):
I had a number of comments, so I just went a head and reworked the dts generation and base flash_map changes.  This removes the need for fa_id and driver match table.

See PR #13081

nvlsianpu commented:
@galak great - lookin into your PR.

nvlsianpu commented:
^^I'm doing this

galak reviewed (dismissed):
Seems like a number of places we need to replace or fix in usb_dfu.c

```
subsys/usb/class/usb_dfu.c:		u8_t bString[USB_BSTRING_LENGTH(FLASH_AREA_IMAGE_0_LABEL)];
subsys/usb/class/usb_dfu.c:		u8_t bString[USB_BSTRING_LENGTH(FLASH_AREA_IMAGE_1_LABEL)];
subsys/usb/class/usb_dfu.c:				FLASH_AREA_IMAGE_0_LABEL),
subsys/usb/class/usb_dfu.c:		.bString = FLASH_AREA_IMAGE_0_LABEL,
subsys/usb/class/usb_dfu.c:				FLASH_AREA_IMAGE_1_LABEL),
subsys/usb/class/usb_dfu.c:		.bString = FLASH_AREA_IMAGE_1_LABEL,
subsys/usb/class/usb_dfu.c:	.flash_upload_size = FLASH_AREA_IMAGE_1_SIZE,
subsys/usb/class/usb_dfu.c:				    FLASH_AREA_IMAGE_0_SIZE;
subsys/usb/class/usb_dfu.c:				    FLASH_AREA_IMAGE_1_SIZE;
```

galak commented:
for mcuboot change, need to fix:
```
tests/subsys/dfu/mcuboot/src/main.c:	for (offs = 0; offs < FLASH_AREA_IMAGE_1_SIZE; offs += sizeof(temp)) {
tests/subsys/dfu/mcuboot/src/main.c:	for (offs = 0; offs < FLASH_AREA_IMAGE_1_SIZE; offs += sizeof(temp)) {
tests/subsys/dfu/mcuboot/src/main.c:	ret = flash_area_read(fa, FLASH_AREA_IMAGE_1_SIZE - sizeof(expectation),
tests/subsys/dfu/mcuboot/src/main.c:	ret = flash_area_read(fa, FLASH_AREA_IMAGE_1_SIZE - sizeof(expectation),
tests/subsys/dfu/mcuboot/src/main.c:	ret = flash_area_read(fa, FLASH_AREA_IMAGE_0_SIZE - sizeof(img_magic),
tests/subsys/dfu/mcuboot/src/main.c:		ret = flash_area_write(fa, FLASH_AREA_IMAGE_0_SIZE - 16,
tests/subsys/dfu/mcuboot/src/main.c:	ret = flash_area_read(fa, FLASH_AREA_IMAGE_0_SIZE - 24, readout,
```

galak commented:
Also:

```
tests/subsys/dfu/img_util/src/main.c:	ret = flash_area_erase(ctx.flash_area, 0, FLASH_AREA_IMAGE_1_SIZE);
```

nvlsianpu committed:
usb/class/usb_dfu: use flash_map data instead of macro labels

Now flash_map is proper place to obtain all information about
flash-areas boundaries.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
^^ @galak I removed all references to macros FLASH_AREA_xxx_SIZE, but FLASH_AREA_IMAGE_xxx_LABEL.

Problem here is that there is no equivalent of this in flash_map now. Anyway macros itself still are present in generated_dts_board_unfixed.h. Can we resolve this in later PR with an USB expert help?

nvlsianpu committed:
usb/class/usb_dfu: use predefined string for image descriptors

Use predefined string for images string descriptors instead
of from DTS build-time generated labels.
This labels will be deprecated.

String content are kept the same as before.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
